[CBR-97] structured logging

CodiePP opened the pull request:
## Description

This PR introduces `katip` as the new logging library, a framework for structured logging (JSON structures). In a first step, we bring in this new library and clean all code from references to the old one. Unfortunately, this translates to changes all over the codebase.

*   introduce `katip` library for logging
*   provide compatible interface
*   logging in IO with `Trace`

replaces PR #3358 which replaced PR #3332 

### notes to reviewers

The order of the commits follows the modules' dependencies.
Start with 'util' where new logging is implemented. This is also the home of `Trace` (by AVieth).
All other modules can be reviewed independently (aka "in parallel").

### notes on `Trace`

Trace has been introduced by @avieth  in `networking` first. It is very handy in that we pass around a `Trace` argument that eventually will `contramap` to the underlying logging. It also allows to stack names to the context using `appendName`, which only affects downstream Trace invocations and is reset after the function returns.
In tests, a simple `noTrace` is passed in as a Trace to inhibit logging.

### notes on `CanJsonLog`

This is a complete implementation of event logging to JSON structures. We will come up with a refactored implementation that is using the new logging and thus should be simpler to maintain. This is not part of this PR. Some parts to the current JSON logging have been commented out, and will be reenabled in a second PR.

## Open issues

to be introduced in subsequent PRs.

* log rotation (CBR-211)
* reenabling logging (CBR-345)
* replace `CanJsonLog` (CBR-321)
* logger tree severity filter (CBR-348)
* from WriterT to StateT (CBR-357)
* remove compatibility layer `CanLog` (CBR-361)

## Linked issues

[CBR-97] U/S
[CBR-207] introduce katip
[CBR-208] rewriting Pos.Util.Log.LogSafe
[CBR-209] 'katip' scribe for JSON logging
[CBR-213] logging in IO - Pos.Util.Trace
[CBR-271] logging configuration parser
[CBR-274] tests for Pos.Util.Log (and Trace)
[CBR-345] cleanup

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [x] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## QA Steps
<!--- Which are the steps needed to test this feature, if any? -->
```
stack build
stack test
```

or in a module, e.g. 'networking':  `stack build cardano-sl-networking`

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

CodiePP commented on scripts/common-functions.sh:
```diff
@@ -247,10 +249,10 @@ function node_cmd {
   # A sloppy test but it'll do for now.
-  topology_first_six_bytes=$(head -c 6 "$topology_file" )
   local topology_first_six_bytes
+  topology_first_six_bytes=$(head -c 6 "$topology_file" )
```
@KtorZ  the 'local' declaration should come before assigning a value.

CodiePP commented on scripts/common-functions.sh:
```diff
@@ -297,7 +299,7 @@ function bench_cmd {
   # the second is the total number of transactions to be sent.
   echo -n " cmd --commands \"send-to-all-genesis $time $time $conc $delay ./tps-sent.csv\""
-  echo -n " --configuration-key bench "
+  echo -n " --configuration-key smallbench "
```
'smallbench' is defined for three "richmen", exactly matching the number of core nodes in this benchmark.
the configuration 'bench' was using seven "richmen", which would allow a the relays and the wallets to create blocks!

CodiePP commented on scripts/bench/runbench.sh:
```diff
@@ -16,7 +16,8 @@
 
-RICH_NODES=3 CONC=4 NUM_TXS="$1" CONFIG_KEY="bench" scripts/launch/demo.sh 6 "$(dirname "$0")/topology" rich_poor
+# nodes wait for three minutes before creating blocks
+system_start=$(($(date +%s) + 60*3)) RICH_NODES=3 CONC=4 NUM_TXS="$1" CONFIG_KEY="smallbench" scripts/launch/demo.sh 6 "$(dirname "$0")/topology" rich_poor
```
it takes time for 'auxx' to precalc some transactions. to limit this idle time, the system start is put three minutes into the future.

CodiePP commented on lib/configuration.yaml:
```diff
@@ -180,6 +180,45 @@ bench: &bench
+        initializer:
+          testBalance:
+            poors:        50000
+            richmen:      3
```
we need a new configuration with exactly three "richmen" (those can create blocks)

erikd commented on util/Pos/Util/Log.hs:
```diff
@@ -0,0 +1,278 @@
+       , LogContext
+       , LogContextT
+       , LoggingHandler
+       -- * Compatibility
```
That is a comment. That is fine.

CodiePP commented on lib/src/Pos/Launcher/Runner.hs:
```diff
@@ -94,26 +91,30 @@ runRealMode pm txpConfig nr@NodeResources {..} act = runServer
+    logTrace' = natTrace liftIO logTrace
     logic :: Logic (RealMode ext)
-    logic = logicFull pm txpConfig ourStakeholderId securityParams jsonLog
+    logic = logicFull logTrace' noTrace pm txpConfig ourStakeholderId securityParams -- TODO jsonLog
```
the 'noTrace' is a placeholder for the JSON logging - in a second PR

CodiePP commented on lib/src/Pos/Launcher/Scenario.hs:
```diff
@@ -93,31 +93,30 @@ runNode' NodeResources {..} workers' plugins' = \diffusion -> do
-            loggerName e
+    reportHandler exc@(SomeException e) =
+        reportOrLogE logTrace
+          (sformat ("Worker/plugin failed with exception: "%shown) e)
```
increase indentation!

CodiePP commented on lib/src/Pos/Util/UserSecret.hs:
```diff
@@ -197,13 +198,6 @@ instance Bi UserSecret where
 
--- | WithLogger is only needed on posix platforms
-#ifdef POSIX
-type MonadMaybeLog m = (MonadIO m, WithLogger m)
```
simplifies on all platforms to 'MonadIO m'

CodiePP commented on lib/src/Pos/WorkMode.hs:
```diff
@@ -134,19 +132,14 @@ instance HasSlogGState (RealModeContext ext) where
-    modifyLoggerName = modifyLoggerNameDefault
-
+-}
+{-
```
will be addressed in CBR-321, just keeping a reference here

CodiePP commented on db/src/Pos/DB/Ssc/Logic/Local.hs:
```diff
@@ -95,22 +98,22 @@ sscNormalize
     bvd <- gsAdoptedBVData
     globalVar <- sscGlobal <$> askSscMem
-    localVar <- sscLocal <$> askSscMem
+    --localVar <- sscLocal <$> askSscMem
```
could be removed.

CodiePP commented on core/src/Pos/Core/Util/LogSafe.hs:
```diff
@@ -71,268 +55,17 @@ module Pos.Core.Util.LogSafe
--- Logging
-----------------------------------------------------------------------------
+-- this is where LogSafe internals are implemented
+import           Pos.Util.Log.LogSafe
```
move closer to logging implementation in 'util'

CodiePP commented on core/src/Pos/Core/Util/LogSafe.hs:
```diff
@@ -71,268 +55,17 @@ module Pos.Core.Util.LogSafe
-import           System.Wlog.LogHandler (LogHandlerTag (HandlerFilelike))
 
-import           Pos.Core (Timestamp)
 import           Pos.Core.Common (Address, Coin)
```
the reason for why we keep this file: provide Buildable instances for types defined outside 'util'

CodiePP commented:
* rebased on current `develop`
   some changes were needed in 'util' because of `RecordWildCard` - I have mentionned the fields explicitely when the structure is not locally defined
* removed the change to turn off optimisations in `wallet/src/Pos/Wallet/Web/Server/Handlers.hs` -> expect Appveyor to fail

erikd commented on core/src/Pos/Core/JsonLog/JsonLogT.hs:
```diff
@@ -68,41 +65,28 @@ instance MonadBaseControl b m => MonadBaseControl b (JsonLogT m) where
-                `catchAny` \e ->
-                    logWarning $ sformat ("can't write json log: "%shown) e
+                `catchAny` \_ -> return ()
+                    --logWarning logTrace $ sformat ("can't write json log: "%shown) e
```
Commented out code.

CodiePP replied:
JSON logging is being done in another PR. to be submitted this week.

CodiePP commented on core/src/Pos/Core/Util/LogSafe.hs:
```diff
@@ -71,268 +55,17 @@ module Pos.Core.Util.LogSafe
-    build = bprint (buildSafeList secure) . getSecureLog
+-- | additional instances for
 
 instance Buildable (SecureLog Text) where
```
it needs access to low-level logging, therefore it was moved to 'util'. these instances remain here to avoid having `katip` as a dependency in other modules than 'util'. also, module 'util' does not have access to these types, defined higher in the module hierarchy.

neongreen commented on auxx/cardano-sl-auxx.cabal:
```diff
@@ -4,8 +4,9 @@ synopsis:            Cardano SL - Auxx
-author:              Serokell
-maintainer:          hi@serokell.io
+homepage:            https://github.com/input-output-hk/cardano-sl
+author:              IOHK
```
When the code has two authors, isn't it customary to keep both of them in the `author` section? :wink:
